feat(whiteframe): add directive to apply md-whiteframe classes

DevVersion opened the pull request:
@EladBezalel You got assigned for this issue, but everywhere is anyone assigned so I just did that PR. If you had a PR in progress, let me know :)

Fixes #6772

DevVersion commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+    var elevation = attr.mdWhiteframe || DEFAULT_DP;
+
+    if (elevation > MAX_DP || elevation < MIN_DP) {
+      $log.warn('md-whiteframe attribute value is invalid. It should be a number between ' + MIN_DP + ' and ' + MAX_DP, element[0]);
```
This needs to be improved?

Frank3K commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+  };
+
+  function postLink(scope, element, attr) {
+    var elevation = attr.mdWhiteframe || DEFAULT_DP;
```
Maybe change to:

``` js
var elevation = parseInt(attr.mdWhiteframe, 10) || DEFAULT_DP;
```

Otherwise values like 1.5 could fail.

DevVersion replied:
Good point, thank you :)

Frank3K commented:
What about the `md-whiteframe-zN` [classes](https://github.com/angular/material/blob/b37ac3adaf821e75a7fffd2efacaf774c6d8e876/src/components/whiteframe/whiteframe.scss)?

DevVersion commented:
That's not said in #6772. But they still can be accessed through the `dp` number. Need more information from @ThomasBurleson if this is desired.

Frank3K commented:
I understand DevVersion. 

When checking older issues I have found that the `-zN` system had been deprecated in favor of the `dpN` system (issues #4706 and #4586). 

Maybe it would be a good time to remove the `-zN` system now. The newly added directive can be introduced at the same time the old `-zN` class-based system is removed. But that's a decision that should be done by the team.

EladBezalel commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+ * @restrict A
+ *
+ * @description
+ * The md-whiteframe directive allows you to apply an elevation shadow to a element.
```
`to an element`

EladBezalel commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+ * @usage
+ * <hljs lang="html">
+ * <div md-whiteframe="3">
+ *   Elevation of 3dp
```
wrap with `<span>`

EladBezalel commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+      elevation = DEFAULT_DP;
+    }
+
+    element.addClass('md-whiteframe-' + elevation + "dp");
```
use single quotes

EladBezalel commented on src/components/whiteframe/whiteframe.spec.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  it('should default to 4dp if no attribute value is specified', function() {
+    var element = buildWhiteframe('');
```
`var element = buildWhiteframe()`

EladBezalel commented:
Besides the comments LGTM, @DevVersion thanks for helping me cleaning my queue :)

DevVersion commented:
@EladBezalel Any time again :) - Applied the suggestions :smile:

ThomasBurleson commented on src/components/whiteframe/demoDirectiveUsage/style.css:
```diff
@@ -0,0 +1,45 @@
+div[md-whiteframe] {
```
*  Attribute selectors are slow in IE.
* This limits the use of `md-whiteframe` to divs ONLY. Should be `md-white-frame`

DevVersion replied:
I thought I should use a div here, as shown in #6772

EladBezalel replied:
I don't know how i missed it, you DON'T want to restrict the use to use only div

DevVersion replied:
Already updated. Just was confused from the description. Sorry

ThomasBurleson commented on src/components/whiteframe/demoDirectiveUsage/style.css:
```diff
@@ -0,0 +1,45 @@
+  margin: 30px;
+  height: 100px;
+}
+@media screen and (max-width: 599px) {
```
Please annotate with breakpoint alias. e.g.

```css
// For breakpoint `-xs`
@media screen and (max-width: 599px) { ... }
```

ThomasBurleson commented on src/components/whiteframe/whiteframe.js:
```diff
@@ -2,4 +2,52 @@
+  function postLink(scope, element, attr) {
+    var elevation = parseInt(attr.mdWhiteframe, 10) || DEFAULT_DP;
+
+    if (elevation > MAX_DP || elevation < MIN_DP) {
```
What about `NaN` ?

DevVersion replied:
Then it will use the default dp, `4`

ThomasBurleson commented on src/components/whiteframe/whiteframe.spec.js:
```diff
@@ -0,0 +1,45 @@
+    return element;
+  }
+
+  it('should default to 4dp if no attribute value is specified', function() {
```
Please tests for `md-whiteframe="nonsense"` or `md-whiteframe="-1"`, etc.

DevVersion replied:
Test for a invalid string is already done.

DevVersion commented:
@ThomasBurleson - Added the missing `-1` test, updated the breakpoints and annotated them and changed attribute selector to `md-whiteframe` (as already used [here](https://github.com/angular/material/blob/master/src/components/whiteframe/demoBasicUsage/style.css#L1))
